File: animal.cs
using System;

namespace AnimalSystem
{
    enum Color
    {
        BLACK, WHITE, GREY, BROWN, DARK_BROWN, BLACK_STRIPES, WHITE_STRIPES
    }

    enum Size
    {
        SMALL, MEDIUM, LARGE, HUGE
    }

    interface IAnimal
    {
        void Sound();
        void Eat();
    }

    interface IMammalia
    {
        void Run();
    }

    interface IDog
    {
        void Bark();
        void Bite();
    }

    interface IAves
    {
        void Fly();
    }

    interface IFish
    {
        void Swimming();
    }

    abstract class Animal : IAnimal
    {
        protected Color color;
        protected Size size;

        public Animal(Color color, Size size)
        {
            this.color = color;
            this.size = size;
        }

        public virtual void Sound()
        {
            Console.WriteLine("animal");
        }

        public virtual void Eat()
        {
            Console.WriteLine("eat animal");
        }

        public virtual string ToStringAnimal()
        {
            return "Animal";
        }
    }

    class Mammalia : Animal, IMammalia
    {
        protected int numberBabies;

        public Mammalia(Color color, Size size, int numberBabies)
            : base(color, size)
        {
            this.numberBabies = numberBabies;
        }

        public override void Sound()
        {
            Console.WriteLine("Mammalia sound");
        }

        public override void Eat()
        {
            Console.WriteLine("Mammalia eat");
        }

        public void Run()
        {
            Console.WriteLine("Mammalia Run");
        }
    }

    class Dog : Mammalia, IDog
    {
        protected bool fierce;

        public Dog(Color color, Size size, int numberBabies, bool fierce)
            : base(color, size, numberBabies)
        {
            this.fierce = fierce;
        }

        public override void Sound()
        {
            Console.WriteLine("sound Dog");
        }

        public override void Eat()
        {
            Console.WriteLine("Dog eat");
        }

        public void Bark()
        {
            Console.WriteLine("Hong!");
        }

        public void Bite()
        {
            Console.WriteLine("kad!!");
        }

        public override string ToStringAnimal()
        {
            return "Dog";
        }
    }

    class ThaiRidgeBack : Dog
    {
        public ThaiRidgeBack(Color color, Size size, int numberBabies, bool fierce)
            : base(color, size, numberBabies, fierce) { }

        public override void Sound()
        {
            Console.WriteLine("Dog thai hong");
        }

        public override void Eat()
        {
            Console.WriteLine("Dog thai eat");
        }
    }

    class Aves : Animal, IAves
    {
        public Aves(Color color, Size size) : base(color, size) { }

        public override void Sound()
        {
            Console.WriteLine("Aves sound");
        }

        public override void Eat()
        {
            Console.WriteLine("Aves eat");
        }

        public void Fly()
        {
            Console.WriteLine("Aves Fly");
        }
    }

    class Bird : Aves
    {
        protected int egg;

        public Bird(Color color, Size size, int egg)
            : base(color, size)
        {
            this.egg = egg;
        }

        public override void Sound()
        {
            Console.WriteLine("Bird sound");
        }

        public override void Eat()
        {
            Console.WriteLine("Bird eat");
        }

        public override string ToStringAnimal()
        {
            return "Bird";
        }
    }

    class HummingBird : Bird
    {
        public HummingBird(Color color, Size size, int egg)
            : base(color, size, egg) { }

        public override void Sound()
        {
            Console.WriteLine("Humming sound");
        }

        public override void Eat()
        {
            Console.WriteLine("humming eat");
        }
    }

    class Osteicthyes : Animal, IFish
    {
        public Osteicthyes(Color color, Size size) : base(color, size) { }

        public override void Sound()
        {
            Console.WriteLine("sound Os");
        }

        public override void Eat()
        {
            Console.WriteLine("eat Os");
        }

        public void Swimming()
        {
            Console.WriteLine("os swimming");
        }
    }

    class Fish : Osteicthyes
    {
        protected string waterGroup;

        public Fish(Color color, Size size, string waterGroup)
            : base(color, size)
        {
            this.waterGroup = waterGroup;
        }

        public override void Sound()
        {
            Console.WriteLine("sound Fish");
        }

        public override void Eat()
        {
            Console.WriteLine("eat Fish");
        }

        public override string ToStringAnimal()
        {
            return "Fish";
        }
    }

    class AngleFish : Fish
    {
        protected string location;

        public AngleFish(Color color, Size size, string waterGroup, string location)
            : base(color, size, waterGroup)
        {
            this.location = location;
        }

        public override void Sound()
        {
            Console.WriteLine("AngleFish Sound");
        }

        public override void Eat()
        {
            Console.WriteLine("Eat AngleFish");
        }

        public override string ToStringAnimal()
        {
            return "AngleFish";
        }
    }
}

File: animal.cpp
#include <iostream>
#include <string>
using namespace std;

enum class Color {
    BLACK, WHITE, GREY, BROWN, DARK_BROWN, BLACK_STRIPES, WHITE_STRIPES
};

enum class Size {
    SMALL, MEDIUM, LARGE, HUGE
};

class Animal_Interface {
public:
    virtual void sound() = 0;
    virtual void eat() = 0;
    virtual ~Animal_Interface() {}
};
class Imamalia{
public:
    virtual void run() = 0;
}
class IDog{
public:
    virtual void bark() =0;
    virtual void bite() =0;
    
};
class IAves{
public:
    virtual fly() = 0;

};
class IFish{
public:
    virtual void swimming() = 0;


};

class Animal:public Animal_Interface {
protected:
    Color color;
    Size size;

public:
    Animal(Color color, Size size) : color(color), size(size) {}
    void sound() override{
        cout<<"animal"<<endl;
    }
    void eat() override{
        cout<<"eat animal"<<endl;
    }
    virtual string toString() {
        return "Animal";
    }

    
};

class Mammalia : public Animal, public Imammalia {
protected:
    int number_Babies;

public:
    Mammalia(Color color, Size size, int number_Babies)
        : Animal(color, size), number_Babies(number_Babies) {}

    void sound() override {
        cout << "Mammalia sound" << endl;
    }

    void eat() override {
        cout << "Mammalia eat" << endl;
    }

    void run() {
        cout << "Mammalia Run" << endl;
    }
};

class Dog : public Mammalia ,public IDog{
protected:
    bool fierce;

public:
    Dog(Color color, Size size, int number_Babies, bool fierce)
        : Mammalia(color, size, number_Babies), fierce(fierce) {}

    void sound()  {
        cout << "sound Dog" << endl;
    }

    void eat()  {
        cout << "Dog eat" << endl;
    }

    void bark() override{
        cout << "Hong!" << endl;
    }

    void Bite() override{
        cout << "kad!!" << endl;
    }

    string toString() override {
        return "Dog";
    }
};

class ThaiRidgeBack : public Dog {
public:
    ThaiRidgeBack(Color color, Size size, int number_Babies, bool fierce)
        : Dog(color, size, number_Babies, fierce) {}

    void sound() override {
        cout << "Dog thai hong" << endl;
    }

    void eat() override {
        cout << "Dog thai eat" << endl;
    }
};

class Aves : public Animal, public IAves{
public:
    Aves(Color color, Size size) : Animal(color, size) {}

    void sound()  {
        cout << "Fly" << endl;
    }

    void eat() {
        cout << "Aves eat" << endl;
    }

    void fly() override{
        cout << "Aves Fly" << endl;
    }
};

class Bird : public Aves {
protected:
    int egg;

public:
    Bird(Color color, Size size, int egg)
        : Aves(color, size), egg(egg) {}

    void sound() override {
        cout << "Bird sound" << endl;
    }

    void eat() override {
        cout << "Bird eat" << endl;
    }

    string toString() override {
        return "Bird";
    }
};

class HummingBird : public Bird {
public:
    HummingBird(Color color, Size size, int egg)
        : Bird(color, size, egg) {}

    void sound() override {
        cout << "Humming sound" << endl;
    }

    void eat() override {
        cout << "humming eat" << endl;
    }
};

class Osteicthyes : public Animal, public IFish{
public:
    Osteicthyes(Color color, Size size) : Animal(color, size) {}

    void sound() override {
        cout << "sound Os" << endl;
    }

    void eat() override {
        cout << "eat Os" << endl;
    }

    void swimming() override{
        cout << "os swimming" << endl;
    }
};

class Fish : public Osteicthyes {
protected:
    string water_Group;

public:
    Fish(Color color, Size size, string water_Group)
        : Osteicthyes(color, size), water_Group(water_Group) {}

    void sound()  {
        cout << "sound Fish" << endl;
    }

    void eat() {
        cout << "eat Fish" << endl;
    }

    string toString() override {
        return "Fish";
    }
};

class AngleFish : public Fish {
protected:
    string location;

public:
    AngleFish(Color color, Size size, string water_Group, string location)
        : Fish(color, size, water_Group), location(location) {}

    void sound() override {
        cout << "AngleFish Sound" << endl;
    }

    void eat() override {
        cout << "Eat AngleFish" << endl;
    }

    string toString() override {
        return "AngleFish";
    }
};

int main() {
    Animal_Interface* a1 = new Dog(Color::BROWN, Size::MEDIUM, 5, true);
    Animal_Interface* a2 = new Bird(Color::WHITE, Size::SMALL, 2);
    Animal_Interface* a3 = new Fish(Color::GREY, Size::LARGE, "Fresh");

    a1->sound();
    a2->sound();
    a3->sound();

    delete a1;
    delete a2;
    delete a3;

    return 0;
}

